File: algorithms/[279]完全平方数/solution.ts
// topics = ["动态规划"]

/**
 * Dynamic Programming
 * time O(n * sqrt(n)), space O(n)
 */
function numSquares(n: number): number {
  const f = new Array<number>(n + 1).fill(0);
  for (let i = 1; i <= n; i += 1) {
    let minn = Number.MAX_VALUE;
    for (let j = 1; j * j <= i; j += 1) {
      minn = Math.min(minn, f[i - j * j]);
    }
    f[i] = minn + 1;
  }
  return f[n];
}

File: algorithms/[279]完全平方数/solution.py
# topics = ["动态规划"]

from sys import maxsize
from math import sqrt


class Solution:
    def numSquares(self, n: int) -> int:
        """
        Dynamic Programming
        time O(n * sqrt(n)), space O(n)
        """
        dp = [maxsize] * (n + 1)
        dp[0] = 0

        for i in range(1, n + 1):
            for j in range(1, int(sqrt(i)) + 1):
                dp[i] = min(dp[i], dp[i - j * j] + 1)

        return dp[n]
